### PR TITLE
Add reading log option for 'Stopped Reading'

### LINKED ISSUE #2016: A Reading Log option for "Stopped Reading"
**Is your feature request related to a problem? Please describe.**
> A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]

https://twitter.com/lethargilistic/status/1110707502495105025

**Describe the solution you'd like**
> A clear and concise description of what you want to happen.

Reading Log drop-down could have an extra option for "Stopped Reading"

**Proposal & Constraints**
> What is the proposed solution / implementation? Is there a precedent of this approach succeeding elsewhere?

It would behave pretty much in the same way as "Finished Reading"

### LINKED ISSUE #4989: Accessing My Reading Logs in development version gives KeyError
Trying to access My Reading Logs in development version gives the error `<class 'KeyError'> at /people/openlibrary/books/want-to-read`

### Traceback:
```
<class 'KeyError'> at /people/openlibrary8967/books/want-to-read
'contact_id'

Python	/openlibrary/openlibrary/core/civicrm.py in <listcomp>, line 82
Web	GET http://localhost:8080/people/openlibrary8967/books/want-to-read

Traceback (innermost first)

/openlibrary/openlibrary/core/civicrm.py in <listcomp>
82.        "contact_id": t.pop("contact_id"), ...

/openlibrary/openlibrary/core/civicrm.py in get_sponsorships_by_contact_id
77.    return [{ ...

/openlibrary/openlibrary/core/sponsorships.py in get_sponsored_editions
45.    return get_sponsorships_by_contact_id(contact_id) if contact_id else [] ...

/openlibrary/openlibrary/plugins/upstream/account.py in GET
745.            sponsorships = get_sponsored_editions(user) ...

/openlibrary/infogami/utils/app.py in delegate
205.        return getattr(cls(), method)(*args) ...

/openlibrary/infogami/utils/app.py in <lambda>
186.    HEAD = GET = POST = PUT = DELETE = lambda self: delegate() ...

/home/openlibrary/.pyenv/versions/3.8.6/lib/python3.8/site-packages/web/application.py in handle_class
494.            return tocall(*args) ...

/home/openlibrary/.pyenv/versions/3.8.6/lib/python3.8/site-packages/web/application.py in _delegate
516.            return handle_class(cls) ...

/home/openlibrary/.pyenv/versions/3.8.6/lib/python3.8/site-packages/web/application.py in handle
270.        return self._delegate(fn, self.fvars, args) ...

/home/openlibrary/.pyenv/versions/3.8.6/lib/python3.8/site-packages/web/application.py in process
279.                    return self.handle() ...
```

### Evidence / Screenshot (if possible)

On adding `?debug=true` at the end of the url:   

### Relevant url

(Development version) http://localhost:8080/people/openlibrary/books/want-to-read

### Steps to Reproduce

1. Set up Open Library locally
2. Visit http://localhost:8080/people/openlibrary/books/want-to-read
3. Visit http://localhost:8080/people/openlibrary/books/want-to-read?debug=true

* Actual: Getting a `<class 'KeyError'> at /people/openlibrary/books/want-to-read` error
* Expected: My actual reading log

### Details

- **Logged in (Y/N)?** Yes
- **Browser type/version?** Brave v 1.22.71 
- **Operating system?** Windows 10
- **Environment (prod/dev/local)?** local

### Proposal & Constraints

This issue is most likely arising either because there doesn't exist a key  `contact_id` inside the [dictionary](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/core/civicrm.py#L78) being returned through the [`get_sponsorships_by_contact_id`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/core/civicrm.py#L53) function, or because the `if contact_id:` statement is not being evaluated properly inside the [`get_contact()`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/core/civicrm.py#L11) function inside [`civicrm.py`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/core/civicrm.py).

### Related files

https://github.com/internetarchive/openlibrary/blob/master/openlibrary/core/civicrm.py

### PR DESCRIPTION
**Pull request by Yashs911**
Closes #2016; Closes #4989 

Added "Stopped Reading" option.
image

### Screenshot

image

### Stakeholders

@mekarpeles

### CODE REVIEW

**Comment by jamesachamp** [comment]
The `ForeignKeyViolation` error is due to the new "Stopped Reading" entry not being present in the database.  After inserting this locally (and fixing the `contact_id` error), I was able to add books to the "Stopped Reading" shelf and view them on the reading log page:

stopped-reading-log-page

In order to insert the new row locally, you will have to attach to the db container and add it there. To do so, first run `docker-compose exec db bash`.  Then, open the Postgres interactive terminal as user openlibrary:
`psql -U openlibrary`
Connect to the database by entering `\c`.
Finally, execute your new insert statement.

Instructions for fixing the `contact_id` error can be found [here](https://github.com/internetarchive/openlibrary/issues/4989#issuecomment-852560508).

File: openlibrary/templates/account/books.html [hunk 1]
```diff
@@ -29,6 +29,9 @@
 $elif key == 'already-read':
   $ og_title = _("Books %(username)s has read", username=userDisplayName)
   $ og_description = _("%(username)s has read %(total)d books. Join %(username)s on OpenLibrary.org and tell the world about the books that you care about.", username=userDisplayName, total=total_items)
+$elif key == 'stopped-reading':
+  $ og_title = _("Books %(username)s have stopped reading", username=userDisplayName)
+  $ og_description = _("%(username)s have stopped reading %(total)d books.", username=userDisplayName, total=total_items)
```
Comment (jamesachamp) [inline openlibrary/templates/account/books.html:34]:
```suggestion
  $ og_title = _("Books %(username)s has stopped reading", username=userDisplayName)
  $ og_description = _("%(username)s has stopped reading %(total)d books.", username=userDisplayName, total=total_items)
```
Small grammar correction.

File: openlibrary/i18n/messages.pot [hunk 2]
```diff
@@ -786,6 +786,7 @@ msgstr ""
 msgid ""
 "This user has chosen to publicly share the books they are <a "
 "href=\"%(username)s/books/currently-reading\">currently reading</a>, <a "
+"href=\"%(username)s/books/already-read\">have stopped reading</a>, <a "
 "href=\"%(username)s/books/already-read\">have already read</a>, and <a "
 "href=\"%(username)s/books/want-to-read\">want to read</a>!"
```
Comment (jamesachamp) [inline openlibrary/i18n/messages.pot:791]: Ensure that this string is updated in the `.po` files as well.

**Comment by Yashs911** [comment]
@jamesachamp I added the row, and for #4989 I added the contact_id, but it showed an error for contribution_status, so I removed it from the dictionary.
The python tests are failing due to flake8 errors, but I don't think we need to break that line.

File: openlibrary/core/civicrm.py [hunk 3]
```diff
@@ -80,6 +80,5 @@ def get_sponsorships_by_contact_id(contact_id=None, isbn=None):
         "context": t.pop(CIVI_CONTEXT),
         "receive_date": t.pop("receive_date"),
         "total_amount": t.pop("total_amount"),
-        "contact_id": t.pop("contact_id"),
-        "contribution_status": t.pop("contribution_status")
+        "contact_id": "270430"
```
Comment (jamesachamp) [inline openlibrary/core/civicrm.py:83]: This is not the correct place to make the change.  There is a method named `fake_civi` in upstream/account.py that mocks API responses for local environments.  That is what needs to be modified.

https://github.com/internetarchive/openlibrary/blob/76f28b83eec379945fff6e2d66a6138adfad180c/openlibrary/plugins/upstream/account.py#L842-L861

**Review by jamesachamp — CHANGES_REQUESTED** [review]
Go ahead and correct the flake8 errors as well.  I'm pretty sure that our CI will prevent deployment if there are failing tests.  Not to mention that code which adheres to a standardized style tends to be easier to maintain.

File: openlibrary/core/civicrm.py (hunk 3 above)
Comment (jamesachamp) [reply openlibrary/core/civicrm.py:83]: When you update this, `Completed` should be a good value for `contribution_status`.

Comment (Yashs911) [reply openlibrary/core/civicrm.py]: I have made the changes.

**Review by jamesachamp — APPROVED** [review]
Thanks @Yashs911!  This looks good to me.

**Comment by jamesachamp** [comment]
**Important:** The following DML must be executed before this code is deployed:

`INSERT INTO bookshelves (name, description) VALUES ('Stopped Reading', 'A list of books I have stopped reading');`

@mekarpeles @cdrini

**Comment by mekarpeles** [comment]
Staff has been struggling to decide whether adding a new "Stopped Reading" shelf ultimately makes sense. For now, I think we are going to close and keep on standby in case there's interest from the community. This PR is very helpful for historical reference as it shows how we may extend our reading log.

**Pull request closed without merging**